File: app/src/main/java/com/letter/days/widget/WidgetSettingActivity.java
package com.letter.days.widget;

import android.appwidget.AppWidgetManager;
import android.content.Intent;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.letter.days.R;
import com.letter.days.anniversary.Anniversary;
import com.letter.days.anniversary.OnAnniItemClickListener;
import com.letter.days.common.SpacesItemDecoration;

import org.litepal.LitePal;

import java.util.List;

public class WidgetSettingActivity extends AppCompatActivity {

    private int widgetId;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setResult(RESULT_CANCELED);

        setContentView(R.layout.activity_widget_setting);
        Toolbar toolbar = findViewById(R.id.Toolbar);
        setSupportActionBar(toolbar);
        this.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle("选择纪念日");
        }

        Intent intent = getIntent();
        widgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID,
                AppWidgetManager.INVALID_APPWIDGET_ID);

        if (widgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish();
            return;
        }

        final List<Anniversary> anniversaryList = LitePal.findAll(Anniversary.class);
        Anniversary closestAnni = new Anniversary();
        closestAnni.setText("最近的纪念日");
        anniversaryList.add(0, closestAnni);

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(new SpacesItemDecoration(1));
        WidgetInfoAdapter widgetInfoAdapter = new WidgetInfoAdapter(anniversaryList);
        widgetInfoAdapter.setOnAnniItemClickListener(new OnAnniItemClickListener() {
            @Override
            public void onItemClick(int position) {
                // TODO: 处理列表点击事件
                WidgetInfo widgetInfo = new WidgetInfo();
                if (position == 0) {
                    widgetInfo.setType(WidgetInfo.WIDGET_TYPE_CLOSEST);
                } else {
                    widgetInfo.setType(WidgetInfo.WIDGET_TYPE_NORMAL);
                    widgetInfo.setAnniId(anniversaryList.get(position).getId());
                }
                widgetInfo.setWidgetId(widgetId);
                widgetInfo.save();
                AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(getApplicationContext());
                AnniversaryWidget.updateAppWidget(getApplicationContext(), appWidgetManager, widgetId);
                Intent resultValue = new Intent();
                resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId);
                setResult(RESULT_OK, resultValue);
                finish();
            }
        });
        recyclerView.setAdapter(widgetInfoAdapter);
    }
}

File: app/src/main/java/com/letter/days/activity/WidgetSettingActivity.kt
package com.letter.days.activity

import android.appwidget.AppWidgetManager
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.*
import androidx.recyclerview.widget.LinearLayoutManager
import com.letter.days.LetterApplication
import com.letter.days.R
import com.letter.days.adapter.BindingViewAdapter
import com.letter.days.databinding.ActivityWidgetSettingBinding
import com.letter.days.viewmodel.DaysListViewModel
import com.letter.presenter.ItemClickPresenter
import com.letter.presenter.ViewPresenter
import com.letter.utils.sendBroadcast
import kotlin.properties.Delegates

class WidgetSettingActivity : BaseActivity(), ItemClickPresenter, ViewPresenter {

    private lateinit var binding: ActivityWidgetSettingBinding
    private val model by lazy {
        ViewModelProvider
            .AndroidViewModelFactory(LetterApplication.instance())
            .create(DaysListViewModel::class.java)
    }

    private var widgetId by Delegates.notNull<Int>();

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityWidgetSettingBinding.inflate(layoutInflater)
        setContentView(binding.root)

        /* 设置Action Bar并使能home按钮 */
        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        setResult(RESULT_CANCELED)
        widgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
        if (widgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish()
            return
        }

        initBinding()
        initModel()
    }

    override fun onResume() {
        super.onResume()
        model.loadDays()
    }

    /**
     * 初始化试图绑定
     */
    private fun initBinding() {
        binding.let {
            it.presenter = this
        }
    }

    /**
     * 初始化view model
     */
    private fun initModel() {
        model.apply {
            daysList.observe(this@WidgetSettingActivity) {
                val adapter = BindingViewAdapter(
                    this@WidgetSettingActivity,
                    R.layout.layout_anniversary_item,
                    it,
                    this@WidgetSettingActivity
                )
                binding.daysListView.apply {
                    this.adapter = adapter
                    layoutManager = LinearLayoutManager(this@WidgetSettingActivity)
                }
            }
        }
    }

    /**
     * 包括widget数据
     * @param widgetId Int widget id
     * @param anniId Int 纪念日id
     */
    private fun saveWidgetInfo(widgetId: Int, anniId: Int) {
        model.saveWidgetInfo(widgetId, anniId) {
            sendBroadcast("android.appwidget.action.APPWIDGET_UPDATE") {
                putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
            }

            val result = Intent()
            result.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, widgetId)
            setResult(RESULT_OK, result)
            finish()
        }
    }

    /**
     * 菜单选项选择处理
     * @param item 被选中的选项
     * @return Boolean 动作是否被处理
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return true
    }

    /**
     * item点击处理
     * @param adapter Any 适配器
     * @param position Int item位置
     */
    override fun onItemClick(adapter: Any, position: Int) {
        saveWidgetInfo(widgetId, model.daysList.value?.get(position)?.id ?: -1)
    }

    /**
     * 视图点击处理
     * @param v View? 视图
     */
    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.closest_text -> {
                saveWidgetInfo(widgetId, -1)
            }
        }
    }
}
